### ISSUE TITLE
Feature request: Option to show cheapest upcoming window

### ISSUE THREAD

**Issue by sas413** [mampfes/ha_epex_spot_sensor#36]
Hello!

At the moment, the sensor only shows the cheapest window on this day. If the window is in the past, no window is displayed.
In many cases it would be nice, to have also displayed the cheapest upcoming window.
Use case: Cheapest window was in the night, I would to like to run the dishwasher on the day a second time. At the moment the sensor gives me no information.

**Comment by LotharWoman** [comment]
Yes, two windows per day is a good option.

`
{% set intervall_1=state_attr('sensor.xxxx', 'raw_today')[0:9] | sort(attribute='value') %}
{% set intervall_2=state_attr('sensor.xxxx', 'raw_today')[9:20] | sort(attribute='value') %}

        {{ (now() >= intervall_1[0].start and now() <= intervall_1[0].end)
        or (now() >= intervall_1[1].start and now() <= intervall_1[1].end)
        or (now() >= intervall_1[2].start and now() <= intervall_1[2].end)
        
        or (now() >= intervall_2[0].start and now() <= intervall_2[0].end)
        or (now() >= intervall_2[1].start and now() <= intervall_2[1].end)
        or (now() >= intervall_2[2].start and now() <= intervall_2[2].end)
        }}
`

In your scenario it is sufficient to allow the time windows to be split? It certainly doesn't always work, but very often.

I think it would be very good if the time windows could also be displayed for the future hours/day and directly under an apex chart of the spot prices.

![grafik](https://github.com/user-attachments/assets/9a0509af-843c-4f8d-917a-404527140827)

**Comment by smoochy** [comment]
Is this integration still being worked on?

I've got some ideas, which would go even beyond sas413 was thinking of. But writing them down would be an effort, which would be useless, if there is no ongoing development anymore.

**Comment by suaveolent** [comment]
Yes this is still on the agenda, but have not yet had time to implement it.

**Comment by smoochy** [comment]
Glad to hear. Sorry if the question sounded a bit harsh, was not intended to be.

So the ideas I had:

Currently I am running a setup described [here](https://github.com/mampfes/ha_epex_spot/discussions/110) and [here](https://github.com/mampfes/ha_epex_spot/discussions/111) for the chart and the appliance schedules and modified it a bit. For instance I have set start times for the laundry, not end times like described in the guide. The Epex spot Data prices are based on the Epex Spot integration and doing some math to convert it into ct/kWh for my needs.

It's looking like this: 

![Image](https://github.com/user-attachments/assets/8ded8101-26de-42d6-8d79-d6e005956b56)

Top are the prices today, bottom are the prices for tomorrow.

I have 4 appliances. 

| Appliance | Duration | Earliest Start - End time |
|--------|--------|--------|
| Dishwasher (Geschirrspüler) | 3h30m | Not needed |
| 30° Wäsche (30 deg laundry) | 3h30m | 4:30am - 11pm |
| 60° Wäsche (60 deg laundry) | 1h40m | 6am - 11pm |
| Brotbackautomat (Breadmachine) | 3h15m | 8am - 11pm | 


So when looking at the daily prices from the chart above:

- For today, the best time to run the dishwasher was from 0:00 to 3:30am. And after that, it is not recommending any time windows anymore for the rest of today and nothing for tomorrow. Obviously this would not work, as with a family the dishwasher is running at least once a day.
- But there would have been plenty of chances to run the 3h30m dishwasher.

I was thinking of:
- Return the cheapest time window for today and tomorrow (as it is now)
- In addition, return a time window, where the appliance would fit in, which is more expensive than the cheapest, but where the price is still x% below the average price for that day. And make x configurable for the user. So I can set 0% below average, or 20% below average price.
- Taking the example, that even if I would have not run the dishwasher in the night from 0:00 to 3:30am, it would still recommend me any time up until the latest at 11:30am to start it so it would be finished by 3pm before the prices hiked. As the price until 3pm was still below the average price of that day.
- And for tomorrow the cheapest window would be from 8:30pm to 12am
- But it could still report another window would be 11:30am to 3pm, as the cost would be below x% below average (I did not do exact math on this, just wanted to mention it as an example)


In addition I was thinking of the following request: Allow the user to configure, during which part of the duration most of the energy is being used. 

Example breadmachine:

![Image](https://github.com/user-attachments/assets/35cb9c56-0dde-4920-8f7a-d0881d3cabe4)

Though it had been started around 6:15pm, most of the energy is not being used in the beginning, where the dough is being treated, but while baking, which is starting around 8:30pm, so in the last hour of the duration.

Example dishwasher:

![Image](https://github.com/user-attachments/assets/59f1c70c-018e-45cd-833c-84ee75d15757)

Most of the energy is being used in the beginning, not so much in between and then in the end when the drying is happening.

So I was thinking of a feature to allow the user to configure several time windows (or at least 1) where most of the energy is being used and that time window is scheduled into the cheapest part of the cheapest window (or the cheapest part of the time window for x% below average price)

It could be configuration for the user like:

For the bread machine:

Duration is set to 3h15m.
User can set energy heave time window from 2h to 3h15m


For the dishwasher:

Duration is set to 3h15m
User can set energy heavy time window from 2h30m to 3h30m.



Not sure if this would also need code changes to the Epex Spot integration.

Obviously I can just look at the chart and make up my own mind for when to schedule any appliance, but hey, we are all lazy and that's why we are here, right? :)


Does this make sense, or do you even have a better idea?

Cheers, 
smoo

**Comment by suaveolent** [comment]
Got it, thanks for clarifying. I am currently trying to sort out some of the issues of the main integration and then come back to this issue.

**Comment by moritzschmitz-oviva** [comment]
I am following this thread with much interest!

@smoochy What are you using to display the graphs on the left side? And perhaps you want to share your templates for the start times as well?

**Comment by smoochy** [comment]
Heya. I followed the guides from [here](https://github.com/mampfes/ha_epex_spot/discussions/110) and [here](https://github.com/mampfes/ha_epex_spot/discussions/111).
